src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -28,12 +28,27 @@ public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchA
             var searchToken = continuationToken as FeedSearchContinuationToken ?? throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
             cancellationToken.ThrowIfCancellationRequested();
 
+            PackageCollectionItem[] installedFeedItems = _installedPackages.GetLatest();
+            PackageCollectionItem[] matchesInstalled = PerformLookup(installedFeedItems, searchToken);
+
             // Remove transitive packages from project references
-            IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
-                .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
+            IEnumerable<PackageCollectionItem> transitivePkgsWithOrigins = _transitivePackages
+                .Where(transitivePkg => transitivePkg
+                    .PackageReferences
+                    .All(pkgRef => pkgRef is ITransitivePackageReferenceContextInfo trPkgRef && trPkgRef.TransitiveOrigins.Any()));
+
+            // First, Search all transitive packages
```

Comment:
Rather than comments that explain exactly what the code is doing in each step, i.e. searching with PerformLookup, then grouping with GroupById, etc., which is easy to determine when we look at the code, maybe add a single comment that explains what is this algorithm doing overall and why.